### ISSUE TITLE
Use JWT header `typ` from RFC 9068 to differentiate between an id_token and access_token

### ISSUE THREAD

**Issue by robotdan** [FusionAuth/fusionauth-issues#1973]
### Description
I don't know if RFC 9068 ever caught on.. so we should look at how many other libraries are platforms support this, but according to that RFC the `typ` claim in the header can be `at+jwt` to indicate this is intended to be used as an access token. 

However, I can't find any reference to this in the OIDC spec, and there is no such recommendation in the OIDC spec that I can find for the `id_token` - which be extension would have the `typ` of `it+jwt` or something like that. 

### Related
- https://www.rfc-editor.org/rfc/rfc9068
- https://openid.net/specs/openid-connect-core-1_0.html

### Community guidelines
All issues filed in this repository must abide by the [FusionAuth community guidelines](https://fusionauth.io/community/forum/topic/1000/code-of-conduct).

**Comment by mooreds** [comment]
Maybe a duplicate of https://github.com/FusionAuth/fusionauth-issues/issues/1455 ?

**Comment by mooreds** [comment]
And 9068 was codified relatively recently, I think in the past year.

**Comment by robotdan** [comment]
I think we will add a claim to make it simpler differentiate the id token from the access token. But I don't think we'll be using 9068. It is just not wide spread enough. 

I think the only way we could support this would be if we built out OAuth profiles and let users opt into various schemes so to speak for how we are going to build out JWTs. 

I have seen more libraries supporting this, such as Spring Security - but it is more of a compatibly statement so that Spring Security doesn't puke when it sees this new value in the `typ` header claim. 

It is not ideal that no industry standard has emerged. I see one IdP called Curity that adds a claim called `purpose` with the value of `id_token` or `access_token`. I don't love the name of the claim, but I appreciate they are doing something. 

It is just sort of funny that everyone says be sure not to accept an id token on an API, but there is not agreed upon mechanism to enforce this. 

Common ideas are to sign the id token with a different key, so your API can't verify the signature, or add a claim a custom claim to the id token. These strategies work for our customers since they can control FusionAuth, but it does require some intentional configuration. 

We will likely add a new claim named `tty` which values of `idt` or `at` so that we can easily document and allow our users to differentiate these token types w/out having a lot of heuristics to decipher the token type.  This will get delivered under another GH issue.

**Comment by mooreds** [comment]
@robotdan is there a reason not to use the 9068 standard, even if it is not widely implemented, rather than use our own custom claim? Would it break backwards compatibility or have a similar bad effect? Maybe I'm missing something, but why couldn't we just document that we are using the `typ` claim?

> I think the only way we could support this would be if we built out OAuth profiles and let users opt into various schemes so to speak for how we are going to build out JWTs.

Another option would be yet another lambda (YAL) to allow for modification of the JWT profile header (or expand the JWT populate lambda to be able to do so).

**Comment by robotdan** [comment]
Yes, the reason is that the JOSE spec says this header `typ` claim is equal to `JWT`.

https://www.rfc-editor.org/rfc/rfc7519#section-5.2

> The "typ" (type) Header Parameter defined by [[JWS](https://www.rfc-editor.org/rfc/rfc7519#ref-JWS)] and [[JWE](https://www.rfc-editor.org/rfc/rfc7519#ref-JWE)] is used
   by JWT applications to declare the media type [[IANA.MediaTypes](https://www.rfc-editor.org/rfc/rfc7519#ref-IANA.MediaTypes)] of
   this complete JWT.  This is intended for use by the JWT application
   when values that are not JWTs could also be present in an application
   data structure that can contain a JWT object; the application can use
   this value to disambiguate among the different kinds of objects that
   might be present.  It will typically not be used by applications when
   it is already known that the object is a JWT.  This parameter is
   ignored by JWT implementations; any processing of this parameter is
   performed by the JWT application.  If present, it is RECOMMENDED that
   its value be "JWT" to indicate that this object is a JWT.  While
   media type names are not case sensitive, it is RECOMMENDED that "JWT"
   always be spelled using uppercase characters for compatibility with
   legacy implementations.  Use of this Header Parameter is OPTIONAL.

All things being equal, I would prefer to use this header `typ` claim value - but from what I can tell, while RFC 9068 has come out of Draft it is still only a proposed standard. So who knows if it will ever become a full standard.
- https://datatracker.ietf.org/doc/rfc9068/

Switching this value based upon a proposed standard seems really risky. 

I do see libraries adding support for this proposed standard. For example - Spring Security added support for this, but it was just to change their validation so they didn't throw out the JWT when the `typ` claim was not equal to `JWT`. So this is an interoperability fix so that if they receive a JWT that has this value, they can tolerate it.

> Currently, if the JWT is having typ as "at+jwt", the token is rejected with message "Failed to authenticate since the JWT was invalid". Spring Security Oauth2 Resource Server with JWT as bearer token should accept typ as "at+jwt" as well.

- https://github.com/spring-projects/spring-security/issues/9900

This makes me think that if we were to change our `typ` header to `at+jwt` or `it+jwt` some integrations will fail if they are attempting to validate this header is equal to something specific such as `JWT` as was the case with Spring Security before the above change. The referenced Spring Security issue is from 2021, so we could assume that all middleware and JWT validation libraries have been updated to allow `at+jwt`. I don't think we can make that assumption.

Also, note that the RFC 9068 doesn't actually prescribe how to designate an Id Token, it only describes `at+jwt` for the access token. So if we add `id+jwt` will anyone else outside of FusionAuth be able to recognize this as a JWT in the form of an Id Token? I don't know the answer, but since the spec doesn't say, it would seem to me that this is open to interpretation. Some may use `id+jwt`, others `idt+jwt` or something else such as using `at+jwt` for access tokens and the traditional `JWT` value for Id Tokens.

I can appreciate why Spring Security is only looking for `at+jwt` because you shouldn't be sending an Id Token to an API - so if they assume their library is protecting APIs that makes sense. However, it is reasonable to assume that third party libraries will also need to validate the Id Token IRL and will they puke if we send a JWT with a header of `type: id+jwt`?

The spec does say 

> This parameter is ignored by JWT implementations; any processing of this parameter is
   performed by the JWT application.  

So in theory that means my argument is invalid, and it is only up to the application to validate this header, but I think the same argument is true which - we have no way to know really if JWT libraries, or applications are asserting this value is equal to `JWT`. And if I use a JWT library in my application, is this library part of my application or is it a JWT process? In reality few application developers are going to write their own application validation code since great off the shelf options exist such as https://github.com/FusionAuth/fusionauth-jwt 😎 But then this becomes part of the application, and then per spec it could or should be processed. Note that FusionAuth JWT doesn't care what this value is - but I don't know what other libraries do.

For these reasons I see making a change like this as very risky unless we add configuration to allow the FusionAuth admin to configure how we should write this header value. This is an option, but it defeats, or partially defeats the reason I want to see this change in FusionAuth, and adding more knobs isn't great as it adds complexity.

> Another option would be yet another lambda (YAL) to allow for modification of the JWT profile header (or expand the JWT populate lambda to be able to do so).

This could help in some cases, but for our internal use cases, we would then have to know which Lambda function mutated the JWT to be able to set our expectations for what to find in the JWT header when we need to validate the token. And since a Lambda function can change over time - this only works if I assume the JWT populate has not been modified for the entire lifetime of a JWT. 

If we were to add a lambda function, we'd then have to default to `JWT` for backwards compatibility, and allow the lambda to modify this claim, but that seems like it adds little value because you would have to opt in to this change to be able to take advantage and it would mean FusionAuth could not take advantage unless a customer turns it on.

**Comment by mooreds** [comment]
I'm happy to have this be closed. It seems like a can of worms with not much value. Also zero upvotes/comments from the community.

One thing worth clearing up, though:

> from what I can tell, while RFC 9068 has come out of Draft it is still only a proposed standard. So who knows if it will ever become a full standard.

The [OAuth 2.0 RFC from 2012 is still a proposed standard](https://datatracker.ietf.org/doc/html/rfc6749). Basically, nothing ever goes to a full internet standard nowadays. 

They [even wrote another RFC about this](https://datatracker.ietf.org/doc/html/rfc7127):

> Initially it was intended that most IETF technical specifications would progress through a series of maturity stages starting with Proposed Standard, then progressing to Draft Standard, then finally to Internet Standard (see Section 6 of RFC 2026). For a number of reasons this progression is not common. Many Proposed Standards are actually deployed on the Internet and used extensively, as stable protocols. This proves the point that the community often deems it unnecessary to upgrade a specification to Internet Standard. Actual practice has been that full progression through the sequence of standards levels is typically quite rare, and most popular IETF protocols remain at Proposed Standard. Over time, the IETF has developed a more extensive review process.Retry

**Comment by robotdan** [comment]
> The [OAuth 2.0 RFC from 2012 is still a proposed standard](https://datatracker.ietf.org/doc/html/rfc6749). Basically, nothing ever goes to a full internet standard nowadays.

Good point. Maybe the better metric is how widespread is the adoption, or how risky the change is knowing there isn't full adoption. 

This one just seems riskier than most because the spec is suggesting a reinterpretation of an existing field, where some OAuth features or extensions even if they never reach full standard status are interoperable because they define new backwards compatible features.

**Comment by juliankrieger** [comment]
What's the proposed solution if I have to work with an IDP that emits `typ: at+jwt`? :(

**Comment by mooreds** [comment]
@juliankrieger this is probably better for a [forum post](https://fusionauth.io/community/forum/) or [opening a support ticket](https://account.fusionauth.io/account/support/) if you have a plan which includes support.
